### ISSUE TITLE
PDF import issues with gofpdi (Failed to read xref: DecodeParms, Invalid memory address)

### ISSUE THREAD

**Issue by rorycl** [rorycl/rm2pdf#4]
Most of the PDF parsing issues with the PDF import library [gofpdi](https://github.com/phpdave11/gofpdi) are resolved according to my latest tests.

The remaining issues are:

* [Failed to read xref table: Failed to read prev xref: Unsupported /DecodeParms](https://github.com/phpdave11/gofpdi/issues/36), and
* [ panic: runtime error: invalid memory address or nil pointer dereference](https://github.com/phpdave11/gofpdi/issues/37)

Hopefully Dave will have a chance to look at these soon.

**Comment by rorycl** [comment]
Although gofpdi has improved, there are still issues.

**Comment by rorycl** [comment]
Some pdf files still cause errors, unfortunately.

**Comment by diegotsutsumi** [comment]
I'm adding another sample of the error with the attached zip file if it helps your debugging.

```
panic: Failed to initialize parser: Failed to read pdf: Failed to read xref table: Unsupported /DecodeParms - only tested with /Columns <= 4 and /Predictor <= 12

goroutine 1 [running]:
github.com/phpdave11/gofpdi.(*Importer).SetSourceStream(0xc0000c5140, 0xc0000c0540)
	/home/tsutsumi/go/pkg/mod/github.com/phpdave11/gofpdi@v1.0.13/importer.go:95 +0x19b
github.com/jung-kurt/gofpdf/contrib/gofpdi.(*Importer).ImportPageFromStream(0x75bba8, {0x647248, 0xc00011d000}, 0x3?, 0x3?, {0x5fe5f8, 0x9})
	/home/tsutsumi/go/pkg/mod/github.com/jung-kurt/gofpdf@v1.16.2/contrib/gofpdi/gofpdi.go:54 +0x45
github.com/jung-kurt/gofpdf/contrib/gofpdi.ImportPageFromStream(...)
	/home/tsutsumi/go/pkg/mod/github.com/jung-kurt/gofpdf@v1.16.2/contrib/gofpdi/gofpdi.go:124
github.com/rorycl/rm2pdf/rmpdf.constructPageWithLayers({0xc0000c0480, 0x1, {0xc0000e6540, 0x2a}, {0x0, 0xed92b2429, 0x75c9a0}, 0x3, 0x3, {0x0, ...}, ...}, ...)
	/home/tsutsumi/remarkable-files/rm2pdf/rmpdf/pdf.go:90 +0x358
github.com/rorycl/rm2pdf/rmpdf.RM2PDF({0x7fff034930a1?, 0x1?}, {0x7fff034930e0, 0x8}, {0x0?, 0xc000030500?}, {0x0, 0x0}, 0x0, {0x0, ...})
	/home/tsutsumi/remarkable-files/rm2pdf/rmpdf/pdf.go:334 +0x3c5
main.main()
	/home/tsutsumi/remarkable-files/rm2pdf/main.go:63 +0x165
```
[DnD_5E_CharacterSheet - Form Fillable copy.zip](https://github.com/rorycl/rm2pdf/files/10737461/DnD_5E_CharacterSheet.-.Form.Fillable.copy.zip)

**Comment by rorycl** [comment]
Hi @diegotsutsumi 

Thanks for the report. Unfortunately this is a gofpdi issue and it seems unlikely the author will be updating the project.

I've tried using the `pdfcpu` library to first process source pdfs, but it often fails too.

The most reliable way I've found to working around the issue of many pdfs having non-standard or corrupt formats is by using the `pdftk` tool outside of `rm2pdf`.

Does resaving the pdf using `pdftk` work for you? For example convert the file with `pdftk <original_file.pdf> cat output <output_file.pdf>`  then move `<output_file.pdf>` to `<original_file.pdf>` (after making a backup of the original file), then re-run `rm2pdf`. This always works for me.

**Comment by diegotsutsumi** [comment]
So yeah, it did work for this case indeed. I see this is documented in `README.md` and in in `./rm2pdf --help` as well, my bad not seeing this before. Thanks for the tip.

It worked for this one, however not for all pdf files I have in my remarkable, `pdftk` doesn't fix a couple of them. I keep seeing `gofpdi` errors for these, unfortunately they're all confidential files I can't upload here.

I guess for these cases we'd be blocked on `gofpdi` fixes, but thanks for pointing the `pdftk` workaround.

**Comment by rorycl** [comment]
Closing as wontfix due to problem with gofpdi; reported upstream.

**Comment by rorycl** [comment]
Closing as wontfix due to problem with gofpdi; reported upstream.
